### PR TITLE
[#5365] Start chatting from the dapp browser 

### LINKED ISSUE #5365: Start chatting from the dapp browser
[comment]: # (Please replace ... with your information. Remove < and >)

### User Story

As a dapp user, I want to chat about the dapp I'm using so that I can learn, get support, and be part of the dapp community.

### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Feature

[comment]: # (Describe the feature you would like, or briefly summarise the bug and what you did, what you expected to happen, and what actually happens. Sections below)
*Summary*:  Status has chat, and it has dapps, but right now they aren't integrated. Let's change that!  We can bring chat to every dapp in status by adding a "chat" icon in the dapp browser and creating an instant community for every dapp.  A user of a dapp like Airswap won't need to hunt around looking for the Airswap community because they can just tap the chat icon and be instantly teleported to the Airswap community.

From our metrics we see that about half of all new users open a dapp, but only half that join a chat. We can leverage the popularity of dapp's in status to drive awareness and engagement with chat.

[Slack context.](https://status-im.slack.com/archives/C5Z5NRYKS/p1533213518000004)

### Solution
From within the dapp browser, a user can tap the chat icon which opens a model and chat UI dedicated to the dapp.

For example, users of Airswap.io can tap the icon and a chat for #airswap.io will open. The user can go back and forth with the dapp and chat seamlessly.

The #airswap.io channel would also be added to the home tab for easy subsequent access.

Concept illustration:

dapp-chat

### PR DESCRIPTION
**Pull request by flexsurfer**
fixes #5365

sep-07-2018 13-07-37

### CODE REVIEW

File: src/status_im/models/browser.cljs [hunk 1]
```diff
@@ -153,7 +153,18 @@
...
+(defn initialize-dapp-permissions
   [{:keys [db all-dapp-permissions]}]
   (let [dapp-permissions (into {} (map #(vector (:dapp %) %) all-dapp-permissions))]
     {:db (assoc db :dapps/permissions dapp-permissions)}))
+
+(defn open-chat-from-browser [host cofx]
+  (let [topic (apply str (map #(let [cc (.charCodeAt % 0)]
+                                 (cond (or (and (> cc 96) (< cc 123))
+                                           (and (> cc 64) (< cc 91))
+                                           (and (> cc 47) (< cc 58)))
```
Comment (goranjovic) [inline src/status_im/models/browser.cljs]: what's behind those numbers? can they be extracted to constants or something more descriptive?

**Review by goranjovic — APPROVED** [review]

File: src/status_im/models/browser.cljs (hunk 1 above)
Comment (flexsurfer) [reply src/status_im/models/browser.cljs]: i'm not sure it's just codes for a-z, A-Z, 0-9

File: src/status_im/chat/events.cljs [hunk 2]
```diff
@@ -134,15 +134,19 @@
 
 (defn- navigate-to-chat
   "Takes coeffects map and chat-id, returns effects necessary for navigation and preloading data"
-  [chat-id {:keys [navigation-replace?]} {:keys [db] :as cofx}]
-  (if navigation-replace?
+  [chat-id {:keys [modal? navigation-replace?]} {:keys [db] :as cofx}]
+  (if modal?
     (handlers-macro/merge-fx cofx
-                             (navigation/replace-view :chat)
+                             {:dispatch [:navigate-to-modal :chat-modal]}
```
Comment (rasom) [inline src/status_im/chat/events.cljs]: Changes related to navigation will cause conflicts with https://github.com/status-im/status-react/pull/5606

File: src/status_im/models/browser.cljs (hunk 1 above)
Comment (rasom) [reply src/status_im/models/browser.cljs]: Can this code be extracted to a separate function at least with some name which will explain what happens actually? It's not clear what is going on here and why.

**Comment by Serhy** [comment]
@flexsurfer, while there are changes requested during review, we tested latest builds from  https://github.com/status-im/status-react/pull/5781#issuecomment-419395497 and there are few points we need to take a look at:

1) Public chat modal window opened from Dapp has the same menu in the `...` top-right menu button like usual public chat that contact can join from `(+)` menu: `Delete chat`, `Clear history` and `Cancel`. Tapping upon `Delete chat` hides the PublicModal window and keeps app unresponsive to actions (on iOS) or shows `Unknown` contact screen (on Android). In any case, both above issues should be "fixed" replacing `Delete chat` menu action in ModalPublicChat by 'Pin chat to Home tab', according to prototype: https://framer.cloud/OFCIZ (also see https://github.com/status-im/status-react/issues/5365#issuecomment-419886013 raised this one)

TF sessions from Android: https://app.testfairy.com/projects/4803622-status/builds/8586691/sessions/4401108975/?accessToken=o3qD7-8THW4OhVIJUgrSQWVPa14

2) The modal window that opens on Android has the top public chat status bar too far from the device menu icons. Also, it overlaps Android status bar so no icons are visible there (time, battery, network statuses, etc...). 
screen shot 2018-09-10 at 14 59 24


3) I've open the public chat modal window from Dapp. When I receive a message in the Modal public chat the message counter increments at top-left while at the time i've seen a message.  See TF session in respect to how should we behave with unread message counter for Modal Public Chat screen https://app.testfairy.com/projects/4803622-status/builds/8586691/sessions/4401109765/?accessToken=ES4-fEPj1oF38SAfeEOqn7Xhpts

4) On IOS when keyboard is opened there is a large empty space between keyboard and messagebox:
screen shot 2018-09-10 at 15 15 16

See TF session: https://app.testfairy.com/projects/4803590-status/builds/8586689/sessions/4401110133/?accessToken=BIVYe0DDe12jqsqMlNzCgryDGuM 


<blockquote><div><strong><a href="https://framer.cloud/OFCIZ">Chat in Dapps</a></strong></div><div>This project was made with Framer, a design tool for turning original concepts into working prototypes.</div></blockquote>
<blockquote><div><strong><a href="https://app.testfairy.com/projects/4803622-status/builds/8586691/sessions/4401108975/?accessToken=o3qD7-8THW4OhVIJUgrSQWVPa14">TestFairy: Status 0.9.28 (7832) @2018</a></strong></div></blockquote>
<blockquote><div><strong><a href="https://app.testfairy.com/projects/4803622-status/builds/8586691/sessions/4401109765/?accessToken=ES4-fEPj1oF38SAfeEOqn7Xhpts">TestFairy: Status 0.9.28 (7832) @2018</a></strong></div></blockquote>
<blockquote><div><strong><a href="https://app.testfairy.com/projects/4803590-status/builds/8586689/sessions/4401110133/?accessToken=BIVYe0DDe12jqsqMlNzCgryDGuM">TestFairy: Status 0.9.28 (7832) @2018</a></strong></div></blockquote>

**Comment by flexsurfer** [comment]
@Serhy i fixed all issues from your comment, about comments in the #5365 issue, let's create for them separate issue, so this is just a MVP

**Comment by Serhy** [comment]
There are a couple of bugs we need to look at:
1) From modal public chat screen opened there an ability to open contact profile screen via tap on it's icon (like in usual public chat). In that case:
Opened on iOS contact profile screen makes an app unresponsible to any action (so the only app shut down could wind back user to use the app)
When open on Android - nothing happens (in fact it seems it opens the screen according to changed device status bar - but the "body" of screen is transparent). Tapping on the area of top-left cross icon closes "transparent" screen and user profile appears. I can then use `Add in contacts`/`Send message`/`Send transaction` actions fine, however if to tap `<` back button - the `TypeError: re-find must match against a string` error pop-up appears crashing the app. 

`Steps to reproduce: (cased on Android)` 
1) Open status, log in to account and navigate to any Dapp (or website e.g. google.com) (Use SimpleTestDapp as the messages exist there already in history)
2) In webview tap Chat icon to open public modal chat
3) Tap contact icon to navigate to it's profile screen.
**Actual result:** No user profile screen is opened
4) Tap on the area where top cross icon displayed
**Actual result:** User profile screen appears
5) Tap `<` button at top-left on user profile screen
TF session with logs a video: https://app.testfairy.com/projects/4803622-status/builds/8592878/sessions/4401483411/?accessToken=k/z3bj/-P8LRTcI3R-X6QS5yLco

As a quick solution, we can we can disable opening profile user screen from public modal chat window upon tap on contact profile icon (user could still do it from respective public chat)


<blockquote><div><strong><a href="https://app.testfairy.com/projects/4803622-status/builds/8592878/sessions/4401483411/?accessToken=k/z3bj/-P8LRTcI3R-X6QS5yLco">TestFairy: Status 0.9.28 (8189) @2018</a></strong></div></blockquote>

**Comment by Serhy** [comment]
2) On iOS if user doesn't have a predictive text enabled - there is a gap between message input box 

screen shot 2018-09-14 at 13 33 43

**Comment by flexsurfer** [comment]
@Serhy fixed 1, can't repro 2 on simulator

**Review by rasom — APPROVED** [review]

**Comment by rasom** [comment]
@Serhy can you check `1.` again please?

**Comment by Serhy** [comment]
yes, https://github.com/status-im/status-react/pull/5781#issuecomment-421311974 is not reproduced (since we disabled navigation to profile username from ModalPublicChat). The rest issues are not reproduced anymore as well (I can't reproduce https://github.com/status-im/status-react/pull/5781#issuecomment-421314113)
There is one issue that, IMO, blocks us from merging it as MVP: 
1) Contact has an opened ModalPublicChat window (e.g. airswap-io) 
2) I received a message in this public channel
3) When user navigates back to Home screen - there is an unread message counter is present on `#airswap-io` public chat that is on Home screen for message contact received while being in ModalPublicChat screen.

https://app.testfairy.com/projects/4803590-status/builds/8597265/sessions/4401774387/?accessToken=UbUKlr9xHUxktQA1AXlm7d4dkg8

@flexsurfer do you think we can synchronise received messages between public chat modal and it's non-modal screens in the scope of this PR?

<blockquote><div><strong><a href="https://app.testfairy.com/projects/4803590-status/builds/8597265/sessions/4401774387/?accessToken=UbUKlr9xHUxktQA1AXlm7d4dkg8">TestFairy: Status 0.9.28 (8518) @2018</a></strong></div></blockquote>

**Comment by flexsurfer** [comment]
@Serhy good catch, fixed

**Comment by asemiankevich** [comment]
the following error is shown when switching networks:

```
00:53 I/ReactNativeJS: NAME: "TypeError"
00:53 I/ReactNativeJS: MESSAGE: undefined is not a function (evaluating 'e.stopWatching')
00:53 I/ReactNativeJS: 
00:53 I/ReactNativeJS: remove_filter_BANG_@index.android.bundle:12:5967186
00:53 I/ReactNativeJS: G__33595_33601@index.android.bundle:12:5975794
00:53 I/ReactNativeJS: index.android.bundle:12:1565532
00:53 I/ReactNativeJS: invoke_interceptor_fn@index.android.bundle:12:1520773
00:53 I/ReactNativeJS: invoke_interceptors@index.android.bundle:12:1521074
00:53 I/ReactNativeJS: execute@index.android.bundle:12:1523444
00:53 I/ReactNativeJS: handle@index.android.bundle:12:1531936
00:53 I/ReactNativeJS: re_frame$router$IEventQueue$_process_1st_event_in_queue$arity$1@index.android.bundle:12:1559021
00:53 I/ReactNativeJS: re_frame$router$IEventQueue$_run_queue$arity$1@index.android.bundle:12:1544006
00:53 I/ReactNativeJS: index.android.bundle:12:1547186
00:53 I/ReactNativeJS: re_frame$router$IEventQueue$_fsm_trigger$arity$3@index.android.bundle:12:1551397
00:53 I/ReactNativeJS: index.android.bundle:12:1545308
00:53 I/ReactNativeJS: index.android.bundle:24:1801
00:53 I/ReactNativeJS: k@index.android.bundle:24:643
00:53 I/ReactNativeJS: q@index.android.bundle:24:1007
00:53 I/ReactNativeJS: callImmediates@index.android.bundle:24:3202
00:53 I/ReactNativeJS: value@index.android.bundle:19:2904
00:53 I/ReactNativeJS: index.android.bundle:19:1377
00:53 I/ReactNativeJS: value@index.android.bundle:19:2621
00:53 I/ReactNativeJS: value@index.android.bundle:19:1347
00:53 I/ReactNativeJS: [native code]
00:53 I/ReactNativeJS: value@index.android.bundle:19:962
00:53 I/ReactNativeJS: [native code]
```

image

STR:
- open status
- create account
- tap profile - development mode - networks - ropsten

After i tap `OK` app crashes

**Pull request merged**